datastream-server/src/main/java/com/linkedin/datastream/server/assignment/LoadBasedPartitionAssigner.java

```diff
@@ -124,4 +139,28 @@
 
     return newAssignments;
   }
+
+  private void validatePartitionCountAndThrow(String datastream, int numTasks, int numPartitions,
+      int maxPartitionsPerTask) {
+    // conversion to long to avoid integer overflow
+    if (numTasks * (long) maxPartitionsPerTask < numPartitions) {
+      String message = String.format("Not enough tasks to fit partitions. Datastream: %s, Number of tasks: %d, " +
+          "number of partitions: %d, max partitions per task: %d", datastream, numTasks, numPartitions,
+          maxPartitionsPerTask);
+      throw new DatastreamRuntimeException(message);
+    }
+  }
+
+  @VisibleForTesting
+  int findTaskWithRoomForAPartition(List<String> tasks, Map<String, Set<String>> partitionMap, int startIndex,
+      int maxPartitionsPerTask) {
+    for (int i = 0; i < tasks.size(); i++) {
+      int currentIndex = (startIndex + i) % tasks.size();
+      String currentTask = tasks.get(currentIndex);
+      if (partitionMap.get(currentTask).size() < maxPartitionsPerTask) {
+        return currentIndex;
+      }
+    }
+    return -1;
```

Comment:
This is not public API and the only caller is LoadBasedPartitionAssigner. I think both returning -1 and throwing an Exception here are viable choices. I'd leave it as is.

Reply:
Just tangent thoughts: In a language which supports passing parameters by reference (*sighs* C#), we would simply update the index integer without having to return anything. In Java the closest thing is wrapping it in an AtomicInteger etc, which is not nice in my opinion.

Reply:
That's true where parameters by reference is supported. Returning -1 means the caller should handle this and that's what is the beauty of Java that it allows you to throw exception so that you don't have to deal with all these corner cases.

Reply:
Again, I'm not opinionated on the above. Since you prefer it that way, I made the change and threw an exception instead of returning -1.